modules/nlu/src/backend/cloud/cache.ts

```diff
@@ -0,0 +1,55 @@
+import { Lock } from 'lock'
+
+export interface Options<T> {
+  getToken?: (res: T) => string
+  getExpiry?: (res: T) => number
+}
+
+const defaultExpiry = 60000 // 10 minutes in ms
+
+export const cache = <T>(authenticate: () => Promise<T>, options?: Options<T>) => {
+  const getToken: (res: any) => string = options?.getToken ?? (res => res)
+  const getExpiry: (res: any) => number = options?.getExpiry ?? (() => defaultExpiry)
+
+  const lock = Lock()
+
+  let token: string | null = null
+  let expiration: number | null = null
+
+  const tokenCache = async () => {
+    if (token && expiration && expiration - Date.now() > 0) {
+      return token
+    }
+
+    return new Promise<string>((resolve, reject) => {
+      lock('cache', unlockFn => {
```

Comment:
\*\*Personnal pref\*\*:

I would probably write a "promised" version of this lib right at the top of this file:

```ts
type UnlockFn = () => {} // not sure about this
const getLock = (key: string): Promise<UnlockFn> => {
    return new Promise(resolve => {
        lock(key, unlockFn => resolve(unlockFn))
    })
}
```

The rest of your real logic code could be callback-free and be written in an `async/await` fashion